base/compiler/tfuncs.jl

```diff
@@ -1619,25 +1619,25 @@ function return_type_tfunc(interp::AbstractInterpreter, argtypes::Vector{Any}, s
                     if isa(rt, Const)
                         # output was computed to be constant
                         return Const(typeof(rt.val))
+                    end
+                    rt = widenconst(rt)
+                    if rt === Bottom || (isconcretetype(rt) && !iskindtype(rt))
+                        # output cannot be improved so it is known for certain
+                        return Const(rt)
+                    elseif !isempty(sv.pclimitations)
```

Comment:
IIUC, this clause contains the only functional change (perhaps modulo some difference between `hasuniquerep(rt)` and `(isconcretetype(rt) && !iskindtype(rt))` that I'm not familiar with in the preceding conditional), adapting this code to the new limitation mechanisms? :)

Reply:
The other condition is arguably more functional: has `hasuniquerep` was the wrong predicate to change to in 88a4db25828, so this changes it back.

Reply:
Awesome, thanks for the explanation! :)